File: python/gbrs/model_io.py
"""
Model serialization functions for GBRS models.

This module provides functions to save and load GBRS models in JSON format,
enabling cross-language compatibility between Python and R implementations.
"""

import json
import numpy as np


def save_model(model, filepath, objective=None, formula=None):
    """
    Save a GBRS model to a JSON file.
    
    Parameters
    ----------
    model : gbrs.GBRS or gbrs.core.Model
        The fitted model object
    filepath : str
        Path to save the model file
    objective : str, optional
        Model objective ('continuous', 'binary', or 'survival')
    formula : str, optional
        Model formula string
    """
    # Handle both GBRS wrapper and raw Model object
    if hasattr(model, "_model"):
        core_model = model._model
    else:
        core_model = model
        
    # Get model parameters
    params = core_model.get_params()
    idxs = core_model.get_idxs()
    split_vals = core_model.get_split_val()
    
    # Build rules list
    rules = []
    for i in range(len(idxs)):
        # Filter out invalid rules if any (though idxs usually valid in trained model)
        # But we might have unused slots if max_n > actual rules?
        # The C++ implementation seems to fill sequentially up to 'i'.
        # But get_idxs returns the full vector of size max_n?
        # Let's check C++ get_idxs. It returns 'this->idxs'.
        # 'this->idxs' is initialized to Zero(max_n).
        # 'this->i' tracks the number of rules.
        # We should probably only save up to 'i'.
        # But we don't have access to 'i' from python directly via get_params?
        # Wait, get_params returns ScoreParams which has 'w'.
        # We can iterate and stop when we hit zeros? No, 0 is valid index.
        # Actually, the C++ prune() method resizes the vectors to 'i'.
        # So if prune() was called (which fit() does), then idxs size should be correct.
        # Let's assume vectors are correct size.
        
        rules.append({
            "idx": int(idxs[i]),
            "split_val": float(split_vals[i]),
            "w": float(params.w[i]),
            "cst": float(params.y0) if i == 0 else 0.0
        })
    
    # Create model dict
    model_dict = {
        "version": "1.0",
        "objective": objective if objective else "unknown",
        "formula": formula if formula else "unknown",
        "rules": rules
    }
    
    # Write to file
    with open(filepath, 'w') as f:
        json.dump(model_dict, f, indent=2)


def load_model(filepath):
    """
    Load a GBRS model from a JSON file.
    
    Parameters
    ----------
    filepath : str
        Path to the model file
        
    Returns
    -------
    gbrs.GBRS
        Loaded GBRS model instance
    """
    with open(filepath, 'r') as f:
        model_dict = json.load(f)
    
    # Validate version
    if model_dict.get("version") != "1.0":
        raise ValueError(f"Unsupported model version: {model_dict.get('version')}")
    
    # Create GBRS instance
    # We don't know the original hyperparameters (n_iter etc), but they don't matter for inference.
    # We can set defaults.
    from gbrs.utils import GBRS
    model = GBRS()
    
    # Set state
    model._set_state(model_dict)
    
    # Store objective/formula if we want to add them to GBRS class later
    # model.objective = model_dict.get("objective")
    # model.formula = model_dict.get("formula")
    
    return model


def save_predictions(predictions, filepath):
    """
    Save model predictions to a JSON file for cross-language comparison.
    
    Parameters
    ----------
    predictions : array-like
        Model predictions
    filepath : str
        Path to save predictions
    """
    # Convert to list for JSON serialization
    pred_list = [float(p) for p in predictions]
    
    with open(filepath, 'w') as f:
        json.dump({"predictions": pred_list}, f, indent=2)


def load_predictions(filepath):
    """
    Load predictions from a JSON file.
    
    Parameters
    ----------
    filepath : str
        Path to predictions file
        
    Returns
    -------
    numpy.ndarray
        Array of predictions
    """
    with open(filepath, 'r') as f:
        data = json.load(f)
    
    return np.array(data["predictions"])

File: R/model_io.R
# Model serialization functions for GBRS models
#
# This file provides functions to save and load GBRS models in JSON format,
# enabling cross-language compatibility between Python and R implementations.

#' Save a GBRS Model to JSON File
#'
#' @param model A gbrs model object
#' @param filepath Path to save the model file
#' @export
save_gbrs <- function(model, filepath) {
  if (!requireNamespace("jsonlite", quietly = TRUE)) stop("jsonlite is required")
  
  # Extract model components
  weights <- model$weights
  
  # Build rules list
  rules <- list()
  for (i in 1:nrow(weights)) {
    rules[[i]] <- list(
      idx = as.integer(weights$idx[i]),
      split_val = as.numeric(weights$split_val[i]),
      w = as.numeric(weights$w[i]),
      cst = as.numeric(weights$cst[i])
    )
  }
  
  # Create model structure
  model_data <- list(
    version = "1.0",
    objective = model$objective,
    formula = deparse(model$formula),
    rules = rules
  )
  
  # Write to JSON file
  jsonlite::write_json(model_data, filepath, pretty = TRUE, auto_unbox = TRUE)
}

#' Load a GBRS Model from JSON File
#'
#' @param filepath Path to the model file
#' @return A gbrs model object
#' @export
load_gbrs <- function(filepath) {
  if (!requireNamespace("jsonlite", quietly = TRUE)) stop("jsonlite is required")
  
  # Read JSON file
  model_data <- jsonlite::read_json(filepath)
  
  # Validate version
  if (model_data$version != "1.0") {
    stop(paste("Unsupported model version:", model_data$version))
  }
  
  # Reconstruct weights data frame
  n_rules <- length(model_data$rules)
  weights <- data.frame(
    idx = integer(n_rules),
    split_val = numeric(n_rules),
    w = numeric(n_rules),
    w1 = numeric(n_rules),
    w2 = numeric(n_rules),
    cst = numeric(n_rules)
  )
  
  for (i in 1:n_rules) {
    rule <- model_data$rules[[i]]
    weights$idx[i] <- rule$idx
    weights$split_val[i] <- rule$split_val
    weights$w[i] <- rule$w
    weights$cst[i] <- rule$cst
    # Note: w1 and w2 are not stored in JSON, set to 0
    weights$w1[i] <- 0
    weights$w2[i] <- 0
  }
  
  # Reconstruct gbrs object
  obj <- list(
    formula = as.formula(model_data$formula),
    weights = weights,
    objective = model_data$objective
  )
  class(obj) <- "gbrs"
  
  return(obj)
}

#' Save Predictions to JSON File
#'
#' @param predictions Numeric vector of predictions
#' @param filepath Path to save predictions
#' @export
save_predictions <- function(predictions, filepath) {
  if (!requireNamespace("jsonlite", quietly = TRUE)) stop("jsonlite is required")
  
  pred_data <- list(predictions = as.numeric(predictions))
  jsonlite::write_json(pred_data, filepath, pretty = TRUE, auto_unbox = FALSE)
}

#' Load Predictions from JSON File
#'
#' @param filepath Path to predictions file
#' @return Numeric vector of predictions
#' @export
load_predictions <- function(filepath) {
  if (!requireNamespace("jsonlite", quietly = TRUE)) stop("jsonlite is required")
  
  pred_data <- jsonlite::read_json(filepath)
  return(as.numeric(pred_data$predictions))
}
